Project structure

HendrikThePendric opened the pull request:
ATM, this is still in draft because I also want to refactor the File component in the same way as I did for the `RadioGroup`. I'll probably get to that on Monday....

For now, if you have time, could you have a look at the following:
1. What do you think of the project structure like this?
1. Your opinion on how I refactored the `RadioGroup`?
1. How about the components in the `./src/for-ui-core` folder, do they make any sense? And do you agree these should be moved to ui-core?
1. Does it make sense to have `./stories/helpers/formDecorator.js`like this?
1. I have stopped exporting `*` from the named root files, but I am exporting `*` from `./src/index.js`. Makes sense?
1. I've started ignoring everything under `./locales` because the files should be auto-generated during a build. Makes sense?
1. There is one quirk reg. the `RadioGroup` component: when the `Field` uses that as its `component` prop, it needs to get `type="select"` (and not "radio"). At first glance this seems weird, but when thinking about it, it actually makes sense: the behaviour of a RadioGroup is identical to a native select.

HendrikThePendric committed:
chore(i18n): gitignore auto-generated i18n files

HendrikThePendric committed:
chore(project-structure): moved files around

HendrikThePendric committed:
fix(file): set button type to "button" to prevent form submitting

HendrikThePendric committed:
chore(form): remove ManagedForm in favour of just using the RFF API

HendrikThePendric committed:
feat(storybook): introduce form decorator

HendrikThePendric committed:
feat(storybook): update createFormDecorator so it accepts all props

HendrikThePendric committed:
refactor: radio-group
- re-implement RadioGroup to be used as a `Field`s `component` prop
- remove Radio component because it will not be used by itself in a form
- extract helper components that contain styling into a separate folder, to be moved to ui-core later
- update RadioGroup stories
- update default onSubmit function in formDecorator
- stub out parsers folder

HendrikThePendric committed:
fix: default onSubmit output

HendrikThePendric committed:
Merge branch 'master' into project-structure

HendrikThePendric commented on stories/RadioGroup.stories.js:
```diff
@@ -10,15 +11,24 @@ const options = [
             validate={required}
+            required
+            options={options}
+            type="select"
```
`type="select"` is perhaps not very intuitive...

Mohammer5 commented on stories/helpers/formDecorator.js:
```diff
@@ -0,0 +1,29 @@
+
+const defaultFormProps = {
+    onSubmit: values => {
+        console.log(
```
No need for `join` here:

```js
[
  '++++++++++++++++',
  'Form was submitted with values:',
  JSON.stringify(values, null, 4),
  '----------------',
].map(console.log)
```

HendrikThePendric replied:
I tried that, but the effect were very different than what I was looking for.... I was joining strings with a newline so I would end up with a nicely formatted console statement. This is what your suggestion produced:
![Screenshot 2019-08-19 at 14 09 44](https://user-images.githubusercontent.com/353236/63264337-1cbe7f80-c28b-11e9-8091-c962004f0e50.png)
perhaps I can tweak it to prevent multiple lines of console.log, but I will need the newlines for the formatting, so I will keep it the way it was.

Mohammer5 replied:
oh my... I mixed up `lodash/fp`'s map and built-in map.. sorry about that

Mohammer5 commented on stories/helpers/formDecorator.js:
```diff
@@ -0,0 +1,29 @@
+}
+
+export const createFormDecorator = formProps => fn => (
+    <Form {...{ ...defaultFormProps, ...formProps }}>
```
Can be:

```js
<Form { ...defaultFormProps } { ...formProps }>
```

Mohammer5 commented on stories/helpers/formDecorator.js:
```diff
@@ -0,0 +1,29 @@
+        {formProps => (
+            <form onSubmit={formProps.handleSubmit}>
+                {fn({ formProps })}
+                <button type="submit">Submit</button>
```
Just a suggestions:
We could use `ui-core`'s `<Button type="button" />` here.

Mohammer5 commented on stories/Conditional.stories.js:
```diff
@@ -1,31 +1,28 @@
+    .addDecorator(createFormDecorator({ mutators: { clear } }))
+    .add('Default', ({ formProps }) => (
+        <>
+            <File name="file1" label="File one" />
```
Suggestions:
We could use a `RadioGroup` here.
Then you don't have to select a file to see the behavior

Mohammer5 commented on src/components/for-ui-core/FlexContainer.js:
```diff
@@ -0,0 +1,69 @@
+import React from 'react'
+import propTypes from 'prop-types'
+
+const FlexContainer = ({
```
I'm very much against styling by props.
It'll cause a separation of styles to two place:
1. css/style files
2. component code

I'd rather have default styles and accept a `className` so styles can be overriden if needed.
It will also allow to use `display: block` when needed (i. e. sometimes you want a completely different behavior on mobile devices)

Furthermore, using a `className` prop is standard in `ui-core`, it'll reduce complexity of this component, will decrease maintenance amount in case of css standards/browser-implementation changes & allows for better customization when deaing with exception-cases.

I don't really mind this here in `ui-forms`, but when moving this one to `ui-core`, we either use `className` or we think about whether we need this component at all, as it's not an atom or molecule but a mere building block.

HendrikThePendric replied:
Managed to get rid of most of these helper components

Mohammer5 reviewed (commented):
Thanks for putting all this work into `ui-forms`! This looks way better than before, I have just a few smaller remarks.
Generally this is going the right direction!

Mohammer5 committed:
test: add enzyme and prepare test setup

Mohammer5 committed:
refactor: rename file parser to reflect file contents

Mohammer5 committed:
test: add test for fileParser

Mohammer5 committed:
test: add test for clear mutator

Mohammer5 commented:
@HendrikThePendric I've added tests for the mutator and parser.
I've also renamed the `file.js` to `fileParser.js` to reflect the exported value

Mohammer5 commented:
I've noticed that after the pre-commit hook was executed, changes to js files were not staged, I assume there's a configuration bug in the linter setup

Mohammer5 commented:
> 1. What do you think of the project structure like this?

I like it!

> 2. Your opinion on how I refactored the RadioGroup?

I think it's fine for now, but once we've implemented the `Radio` as a standalone input, there's no need for this one anymore

> 3. How about the components in the ./src/for-ui-core folder, do they make any sense? And do you agree these should be moved to ui-core?

Yes an no, see my comments regarding the `FlexContainer`

> 4. Does it make sense to have ./stories/helpers/formDecorator.jslike this?

Absolutely yes

> 5. I have stopped exporting * from the named root files, but I am exporting * from ./src/index.js. Makes sense?

We could move all the exports from the `mutators.js`, `validators.js` and `inputs.js` to the `src/index.js` actually, just like we do it in `ui-core` :thinking: 

> 6. I've started ignoring everything under ./locales because the files should be auto-generated 
 during a build. Makes sense?

Yes

> 7. There is one quirk reg. the RadioGroup component: when the Field uses that as its component prop, it needs to get type="select" (and not "radio"). At first glance this seems weird, but when thinking about it, it actually makes sense: the behaviour of a RadioGroup is identical to a native select.

Doesn't really matter as we'll get rid of this component anyways. See my 2nd answer

HendrikThePendric committed:
fix(radio-group): spread remaining props onto Radios

HendrikThePendric committed:
refactor(radio-group): separate ui-core from ui-form component

HendrikThePendric committed:
refactor: file
- split file into a component for ui-forms and ui-core
- remove mutators because they are not in use right now
- introduce FormControl for wrapping and spacing
- based file on existing button

HendrikThePendric committed:
fix(radio-group): extract Help from ui-core component

HendrikThePendric committed:
fix(field): change incorrect meta.data prop-type

HendrikThePendric committed:
chore(conditional): use RadioGroup in story to toggle Conditional

HendrikThePendric committed:
chore(exports): just export from a single file

HendrikThePendric commented:
Just marked this "Ready for review" a moment ago, and then realised this is not strictly true. The components that are currently in the `for-ui-core` folder would have to be lifted out, put into ui-core, and then we'd have to upgrade the ui-core version here and start using them. So please don't merge yet....

Also, I think I have done a bit too much in this branch, and as a result we run the risk that this is going to take very long to perfect and finally merge in... Perhaps we could agree to just focus on the project structure and the way that these components are structured. For module-specific issues/features we can create new (JIRA) issues.

HendrikThePendric committed:
Merge branch 'master' into project-structure

HendrikThePendric committed:
chore(deps): switch to File branch of ui-core to use future components

HendrikThePendric committed:
chore: remove redundant files

HendrikThePendric committed:
feat(field-adapter): introduce component that maps FF props to ui-core

HendrikThePendric committed:
refactor(radio-group): compose from FieldSet, Legend and FieldAdapter

HendrikThePendric committed:
fix: adapt FileInput and RadioGroup to new structure in ui-core

HendrikThePendric committed:
docs(field): list props that are "swallowed" by `Field` in readme

HendrikThePendric committed:
fix(file-input): prevent duplicate files and add multi-file support

HendrikThePendric committed:
chore(deps): switch to ui-core next branch because FileInput is there

HendrikThePendric committed:
chore(file-input): rename story file to match component name

HendrikThePendric committed:
docs: updated readme with known issues

HendrikThePendric commented:
@Mohammer5 I think we should merge this into master now. It's not perfect, but the structure is clear and remaining issues can be addressed in separate branches.
